### DISCUSSION TITLE
Allow yield to return type by convention

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/csharplang#4726] [Language Ideas]
Currently yield requires you to return IEnumerable or IEnumerator.
It is not possible to return a derived Interface or return my own type.

Use case, 
interface IRefEnumerable<T> { ref T Current{get;} bool MoveNext(); }

I could yield return with this type if it was allowed.

**Comment by YairHalberstadt** [comment]
How would the compiler know how to generate code for these types?
